extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java

```diff
@@ -506,29 +507,105 @@ public void handle() throws InterruptedException, ExecutionException, TimeoutExc
+  void resetInternal(DataSourceMetadata dataSourceMetadata)
   {
-    boolean result = indexerMetadataStorageCoordinator.deleteDataSourceMetadata(dataSource);
-    log.info("Reset dataSource[%s] - dataSource metadata entry deleted? [%s]", dataSource, result);
+    if (dataSourceMetadata == null) {
+      // Reset everything
+      boolean result = indexerMetadataStorageCoordinator.deleteDataSourceMetadata(dataSource);
+      log.info("Reset dataSource[%s] - dataSource metadata entry deleted? [%s]", dataSource, result);
+      killTaskGroupForPartitions(JavaCompatUtils.keySet(taskGroups));
+    } else {
+      // Reset only the partitions in dataSourceMetadata if it has not been reset yet
+      final KafkaDataSourceMetadata resetKafkaMetadata = (KafkaDataSourceMetadata) dataSourceMetadata;
+
+      if (resetKafkaMetadata.getKafkaPartitions().getTopic().equals(ioConfig.getTopic())) {
+        // currentMetadata can be null
+        final KafkaDataSourceMetadata currentMetadata = (KafkaDataSourceMetadata) indexerMetadataStorageCoordinator.getDataSourceMetadata(
+            dataSource);
+
+        // defend against consecutive reset requests from replicas
+        // as well as the case where the metadata store do not have an entry for the reset partitions
+        boolean doReset = false;
+        for (Map.Entry<Integer, Long> resetPartitionOffset : resetKafkaMetadata.getKafkaPartitions()
+                                                                               .getPartitionOffsetMap()
+                                                                               .entrySet()) {
+          final Long partitionOffsetInMetadataStore = currentMetadata == null
+                                                      ? null
+                                                      : currentMetadata.getKafkaPartitions()
+                                                                       .getPartitionOffsetMap()
+                                                                       .get(resetPartitionOffset.getKey());
+          final TaskGroup partitionTaskGroup = taskGroups.get(getTaskGroupIdForPartition(resetPartitionOffset.getKey()));
+          if (partitionOffsetInMetadataStore != null ||
+              (partitionTaskGroup != null && partitionTaskGroup.partitionOffsets.get(resetPartitionOffset.getKey())
+                                                                                .equals(resetPartitionOffset.getValue()))) {
+            doReset = true;
+            break;
+          }
+        }
 
-    for (TaskGroup taskGroup : taskGroups.values()) {
-      for (Map.Entry<String, TaskData> entry : taskGroup.tasks.entrySet()) {
-        String taskId = entry.getKey();
-        log.info("Reset dataSource[%s] - killing task [%s]", dataSource, taskId);
-        killTask(taskId);
+        if (!doReset) {
+          return;
+        }
+
+        boolean metadataUpdateSuccess = false;
+        if (currentMetadata == null) {
```

Comment:
`currentMetadata` can be null. In case metadata is null, it is still possible for running tasks to get stuck and `doReset` will be true as the condition on line 555 `partitionTaskGroup != null && partitionTaskGroup.partitionOffsets.get(resetPartitionOffset.getKey()).equals(resetPartitionOffset.getValue())` will be true